File: frontend/src/api/pricing.js
import http from './http'

export async function getLatestDealPrice(params) {
  const { customer_id, product_id } = params || {}
  if (!customer_id || !product_id) return null

  try {
    const res = await http.get('/api/pricing/last-price', { params: { customer_id, product_id } })
    return res.data || null
  } catch (error) {
    return null
  }
}

export async function getPricingHistory(params) {
  const { customer_id, product_id, page = 1, page_size = 20, start_date, end_date } = params || {}
  if (!customer_id || !product_id) return { items: [], meta: { total: 0, page: 1, page_size: 20, pages: 0 } }

  try {
    const res = await http.get('/api/pricing/history', {
      params: { customer_id, product_id, page, page_size, start_date, end_date }
    })
    return res.data
  } catch (error) {
    return { items: [], meta: { total: 0, page: 1, page_size: 20, pages: 0 } }
  }
}

export const getLastPricing = getLatestDealPrice
export const getLastPrice = getLatestDealPrice

File: backend/app/routers/pricing.py
from fastapi import APIRouter, Depends, Query
from sqlmodel import Session

from app.db.session import get_session
from app.schemas.pricing import PricingHistoryResponse, PricingLastResponse, ProductTrendItem
from app.services import pricing_service

router = APIRouter(prefix="/api/pricing", tags=["Pricing/History"])


@router.get("/last-price", response_model=PricingLastResponse)
def get_last_price(
    customer_id: int = Query(..., ge=1),
    product_id: int = Query(..., ge=1),
    session: Session = Depends(get_session),
):
    return pricing_service.last_pricing(session, customer_id, product_id)


@router.get("/history", response_model=PricingHistoryResponse)
def get_history(
    customer_id: int = Query(..., ge=1),
    product_id: int = Query(..., ge=1),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=200),
    start_date: str | None = Query(None),
    end_date: str | None = Query(None),
    session: Session = Depends(get_session),
):
    return pricing_service.pricing_history(
        session,
        customer_id,
        product_id,
        page=page,
        page_size=page_size,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/product_trend", response_model=list[ProductTrendItem])
def get_product_trend(
    product_id: int = Query(..., ge=1),
    limit: int = Query(50, ge=1, le=200),
    session: Session = Depends(get_session),
):
    return pricing_service.product_trend(session, product_id, limit=limit)
